core/src/main/scala/akka/persistence/r2dbc/internal/BySliceQuery.scala

```diff
@@ -70,6 +88,12 @@ import org.slf4j.Logger
     }
   }
 
+  // only filter by highest seen seq nr when the next query is the same timestamp (or when unknown for initial queries)
+  private def highestSeenSeqNr(previous: TimestampOffset, latest: TimestampOffset): Option[Long] =
+    Option.when((previous == TimestampOffset.Zero || previous.timestamp == latest.timestamp) && latest.seen.nonEmpty) {
+      latest.seen.values.max
```

Comment:
> Note that sequence numbers are per persistence id, so a later timestamp can have an earlier sequence number

That is correct. How is that handled? I guess it is caught by backtracking in the same way when visibility of pid-A may be seen before another pid-B even though timestamp of pid-B is before pid-A.

Reply:
If the events are there, it handles it by only filtering by highest sequence number for the same timestamp as the query starting timestamp. See also where the query is adjusted, and the comment there.

So if there are events (ordered by timestamp, seq nr):

timestamp | seq nr | pid  | event
----------|--------|------|------
t1        | 1      | pid1 | A1
t1        | 1      | pid2 | B1
t1        | 1      | pid3 | C1
t1        | 2      | pid1 | A2
t1        | 2      | pid2 | B2
t1        | 3      | pid1 | A3
t2        | 2      | pid3 | C2
t2        | 3      | pid2 | B3
t2        | 4      | pid1 | A4

And a buffer size of 4. First query only gets to the 4th event (A2). Before it would restart from `timestamp >= t1`, stuck on repeating the same 4 events (A1, B1, C1, A2). Now it will use the query `timestamp >= t1 AND (timestamp != t1 OR seq_nr >= 2)` for the next query, starting from the 4th event now, processing (A2 (deduplicated), B2, A3, C2). The seq number filter only applies to the starting timestamp, so that the query is otherwise just by timestamp as usual.

Applies for the both the regular queries and the backtracking queries. Backtracking catching any late arriving events as usual.

Reply:
Good, that should work fine. What I was thinking about would be classified as late arriving event.